src/main/java/com/yodle/vantage/component/service/ComponentService.java

```diff
@@ -81,45 +85,143 @@ public Version createOrUpdateVersion(Version version) {
+                        .map(dep -> dep.getVersion().toId())
+                        .collect(Collectors.toSet())
+        );
 
-    private void saveRequestedDependencies(Version parent, Set<VersionId> componentsWithVersionsCreated) {
-        for (Dependency dep : parent.getRequestedDependencies()) {
-            componentDao.ensureCreated(dep.getVersion().getComponent());
-            if (versionDao.createRequestedImplicitDependency(parent, dep.getVersion(), dep.getProfiles())) {
-                componentsWithVersionsCreated.add(dep.getVersion().toId());
+        if (!excludeRequestedDependencies) {
+            shadowVersionsToEnsureCreated.addAll(
+                    version.getResolvedDependencies().stream()
+                            .flatMap(dep -> dep.getVersion().getRequestedDependencies().stream())
+                            .filter(dep -> !versionPurifier.isRealVersion(dep.getVersion().getVersion()))
+                            .map(dep -> dep.getVersion().toId())
+                            .map(versionPurifier::purifyVersion)
+                            .collect(Collectors.toSet())
+            );
+
+            shadowVersionsToEnsureCreated.addAll(
+                    version.getRequestedDependencies().stream()
+                            .filter(dep -> !versionPurifier.isRealVersion(dep.getVersion().getVersion()))
+                            .map(dep -> dep.getVersion().toId())
+                            .map(versionPurifier::purifyVersion)
+                            .collect(Collectors.toSet())
+            );
+        }
+
+        //Regardless of versioning scheme, we're doing maven ordering.  Maven versions need to be done in maven ordering
+        //since when we fix precedence we want to fix precedence in maven order to prevent deadlocks.  Non-maven version
+        // precedence is based on create
+        //timestamp, which means that if we were ever fixing precedence for multiple non-maven versions for the same component
+        //at the same time, it means we created two non-maven versions at the same time and we don't have an ordering
+        //guarantee between them so we can insert them in the precedence list in an arbitrary order.  We choose maven
+        //ordering to ensure we create them in a deterministic order so as to prevent deadlocks
+        Comparator<VersionId> versionComparator = (o1, o2) -> {
+            if (o1.getComponent().equals(o2.getComponent())) {
+                return new ComparableVersion(o1.getVersion()).compareTo(new ComparableVersion(o2.getVersion()));
+            } else {
+                return o1.getComponent().compareTo(o2.getComponent());
             }
+
+        };
+
+        shadowVersionsToEnsureCreated
+                .stream()
+                .sorted(versionComparator)
+                .forEach(v -> versionDao.createShadowVersion(v.getComponent(), v.getVersion()));
+
+        return realVersionsToEnsureCreated
```

Comment:
this method returns the versions we actually created (as opposed to matched) so we know to fix precedence for those versions later